src/tree_annotation/database.cljs

```diff
@@ -25,172 +27,323 @@
+    (f node)
+    (let [children (update (:children node) (first index) update-node f (rest index))]
+      (assoc node :children children))))
 
-;------------------------;
-; Node children requests ;
-;------------------------;
+(defn update-forest [forest f index]
+  "Updates the node at `index` in `forest` by applying `f` to it.
+Returns the updated forest."
+  (if (empty? index)
+    forest
+    (update forest (first index) update-node f (rest index))))
 
-(defn get-children-coords [coord]
-  (get-in @db [:nodes coord :children-coords]))
-
-(defn leaf? [coord]
-  (empty? (get-children-coords coord)))
-
-(defn last-leaf? [coord]
-  (and (leaf? coord)
-       (= (first coord)
-          (->> (get-node-coords)
-               (map first)
-               (apply max)))))
-
-(defn last-leaf-or-inner-node? [coord]
-  (or (last-leaf? coord)
-      (not (leaf? coord))))
-
-;----------------------;
-; Node parent requests ;
-;----------------------;
-
-(defn get-parent-coord [coord]
-  (get-in @db [:nodes coord :parent-coord]))
-
-(defn set-parent-coord [coord parent-coord]
-  (swap! db assoc-in [:nodes coord :parent-coord] parent-coord))
-
-(defn del-parent-coord [coord]
-  (swap! db
-    (fn [db]
-      (assoc-in db [:nodes coord]
-        (dissoc (get-in db [:nodes coord]) :parent-coord)))))
-
-(declare del-node)
-
-(defn del-ancestors [coord]
-  (let [parent-coord (get-parent-coord coord)]
-    (when (some? parent-coord)
-      (del-ancestors parent-coord)
-      (del-node parent-coord)
-      (del-parent-coord coord))))
-
-;------------------------------;
-; Node add and delete requests ;
-;------------------------------;
-
-(defn add-node [{:keys [x y length label children-coords state]}]
-  (let [properties {:length length
-                    :label label
-                    :children-coords children-coords
-                    :state state}]
-    (swap! db assoc-in [:nodes [x y]] properties)
-    (doall (map #(set-parent-coord % [x y]) children-coords))))
-
-(defn del-all-nodes []
-  (swap! db assoc :nodes {}))
-
-(defn del-node [coord]
-  (when (last-leaf-or-inner-node? coord)
-    ; delete all ancestors of the node (parent, grand-parent, ...)
-    (del-ancestors coord)
-    ; delete node coord as the parent of its children
-    (doall 
-      (for [child-coord (get-children-coords coord)]
-        (swap! db
-          (fn [db]
-            (assoc-in db [:nodes child-coord]
-              (dissoc (get-in db [:nodes child-coord]) :parent-coord))))))
-    ; delete the node itself
-    (swap! db 
-      (fn [db]
-        (assoc db :nodes 
-          (dissoc (:nodes db) coord))))))
-
-(defn del-selected-nodes []
-  (doall (for [coord (get-selected-node-coords)]
-           (del-node coord))))
+(defn map-tree [f tree]
+  "Map `f` over all nodes in `tree`.
+`f` is first applied to children and then to the parent."
+  (let [children' (mapv (partial map-tree f) (:children tree))
+        tree' (assoc tree :children children')]
+    (f tree)))
```

Comment:
It's ok that `f` has access to the children, because it might make use of properties of the children, e.g., when recomputing coordinates. The docstring even mentions that `f` is applied first to the children and then to the parent. However, the implementation was incorrect and used the old node, but this is fixed now.

Reply:
mh, `map-tree` is not used at all. Removed.